src/Http/Paginators/RequestPaginator.php

```diff
@@ -0,0 +1,152 @@
+//           and have that callback set the next 'page' on the new Request?
+
+/**
+ * @template TRequest of \Saloon\Contracts\Request
+ * @template TResponse of \Saloon\Contracts\Response
+ *
+ * @implements RequestPaginatorContract<TRequest, TResponse>
+ */
+abstract class RequestPaginator implements RequestPaginatorContract, SerialisableRequestPaginator
+{
+    /**
+     * @var bool
+     */
+    protected bool $async = false;
+
+    /**
+     * @var TResponse|null
+     */
+    protected Response|null $lastResponse = null;
+
+    /**
+     * @param \Saloon\Contracts\Connector $connector
+     * @param TRequest $originalRequest
+     * @param int|null $limit
+     */
+    public function __construct(
+        protected readonly Connector $connector,
+        protected readonly Request $originalRequest,
+        protected readonly int|null $limit,
+    ) {}
+
+    /**
+     * @param callable(int $pendingRequests): (int)|int $concurrency
+     * @param callable(TResponse $response, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $responseHandler
+     * @param callable(mixed $reason, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $exceptionHandler
+     * @return \Saloon\Contracts\Pool
+     */
+    public function pool(
+        callable|int $concurrency = 5,
+        callable|null $responseHandler = null,
+        callable|null $exceptionHandler = null,
+    ): Pool {
+        // The purpose of a pool is to concurrently send requests.
+        // So 'force' set async.
+        $this->async();
+
+        return $this->connector->pool(
+            $this,
+            $concurrency,
+            $responseHandler,
+            $exceptionHandler,
+        );
+    }
+
+    /**
+     * @param bool $async
+     *
+     * @return $this
+     */
+    public function async(bool $async = true): static
+    {
+        $this->async = $async;
+
+        return $this;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isAsync(): bool
+    {
+        return $this->async;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function limit(): ?int
+    {
+        return $this->limit;
+    }
+
+    /**
+     * @return \Saloon\Contracts\Response|null
+     */
+    public function lastResponse(): ?Response
+    {
+        return $this->lastResponse;
+    }
+
+    /**
+     * @return void
+     */
+    public function rewind(): void
+    {
+        // Nullify the last response, so we don't accidentally check if there are more pages, even though we're starting over.
+        $this->lastResponse = null;
+    }
+
+    /**
+     * @return array{
+     *     connector: \Saloon\Contracts\Connector,
+     *     original_request: \Saloon\Contracts\Request,
+     *     limit: int|null,
+     * }
+     */
+    public function __serialize(): array
+    {
+        // TODO: figure out how to serialise the 'has next page' resolver/callback.
+
```

Comment:
If we switched to making the user extend their own class, we wouldn't have to worry about serializing closures

Reply:
I have added the `hasNextPage()` method, which needs to be implemented.
There's a TODO on it, to figure out a better name, though.
`isFinished()`? Or something like it, to better fit offset-based paging.

Reply:
isFinished works!